game-core/src/main/java/games/strategy/triplea/ai/pro/data/ProTerritoryManager.java

```diff
@@ -360,103 +359,33 @@ private static void findScrambleOptions(
-            .and(
-                Matches.territoryHasUnitsThatMatch(
-                    Matches.unitCanScramble()
-                        .and(Matches.unitIsEnemyOf(data, player))
-                        .and(Matches.unitIsNotDisabled())))
-            .and(Matches.territoryHasUnitsThatMatch(airbasesCanScramble))
-            .andIf(fromIslandOnly, Matches.territoryIsIsland())
-            .build();
-
-    // Find potential territories to scramble from
-    final Map<Territory, Set<Territory>> scrambleTerrs = new HashMap<>();
-    for (final Territory t : moveMap.keySet()) {
-      if (t.isWater() || !toSeaOnly) {
-        final Set<Territory> canScrambleFrom =
-            new HashSet<>(
-                CollectionUtils.getMatches(
-                    data.getMap().getNeighbors(t, maxScrambleDistance), canScramble));
-        if (!canScrambleFrom.isEmpty()) {
-          scrambleTerrs.put(t, canScrambleFrom);
-        }
-      }
-    }
-    if (scrambleTerrs.isEmpty()) {
-      return;
-    }
-
-    // Find potential max units that can be scrambled to each territory
-    for (final Territory to : scrambleTerrs.keySet()) {
-      for (final Territory from : scrambleTerrs.get(to)) {
-
-        // Find potential scramble units from territory
-        final Collection<Unit> airbases = from.getUnitCollection().getMatches(airbasesCanScramble);
-        final int maxCanScramble = getMaxScrambleCount(airbases);
-        final Route toBattleRoute =
-            data.getMap().getRoute(from, to, Matches.territoryIsNotImpassable());
-        List<Unit> canScrambleAir =
-            from.getUnitCollection()
-                .getMatches(
-                    Matches.unitIsEnemyOf(data, player)
-                        .and(Matches.unitCanScramble())
-                        .and(Matches.unitIsNotDisabled())
-                        .and(Matches.unitWasScrambled().negate())
-                        .and(Matches.unitCanScrambleOnRouteDistance(toBattleRoute)));
-
-        // Add max scramble units
-        if (maxCanScramble > 0 && !canScrambleAir.isEmpty()) {
-          if (maxCanScramble < canScrambleAir.size()) {
-            canScrambleAir.sort(
-                Comparator.<Unit>comparingDouble(
-                        o ->
-                            ProBattleUtils.estimateStrength(
-                                to, List.of(o), new ArrayList<>(), false))
-                    .reversed());
-            canScrambleAir = canScrambleAir.subList(0, maxCanScramble);
-          }
-          moveMap.get(to).getMaxScrambleUnits().addAll(canScrambleAir);
+    final var scrambleLogic = new ScrambleLogic(data, player, moveMap.keySet());
+    for (final var territoryToScramblersEntry :
+        scrambleLogic.getUnitsThatCanScrambleByDestination().entrySet()) {
+      final Territory to = territoryToScramblersEntry.getKey();
+      for (final var fromToScrambers : territoryToScramblersEntry.getValue().entrySet()) {
```

Comment:
Would it be better to specify the type here instead of var since its not very obvious?

Reply:
Done and switched iteration to be on values.